File: server/src/dlp.rs
use regex::Regex;
use std::sync::LazyLock;

// Expressões Regulares de Prevenção de Fuga de Dados (DLP)
static CREDIT_CARD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?:\d[ -]*?){13,16}").unwrap());

static NIF_RE: LazyLock<Regex> = LazyLock::new(|| {
    // Valida padrões parecidos com o NIF Português (9 dígitos)
    Regex::new(r"\b[1-9]\d{8}\b").unwrap()
});

static API_KEY_RE: LazyLock<Regex> = LazyLock::new(|| {
    // Simula uma chave de API (ex: sk-abcdef1234567890abcdef1234567890)
    Regex::new(r"sk-[a-zA-Z0-9]{32}").unwrap()
});

/// Palavras ofensivas (PT + EN) a mascarar nas legendas/transcrição — lista
/// deliberadamente curta e conservadora (só termos claramente ofensivos), para
/// não censurar palavras legítimas. `(?i)` = case-insensitive; `\b` = fronteira
/// de palavra (não apanha substrings inocentes). Ajustável por org no futuro.
static PROFANITY_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"(?i)\b(?:merda|caralho|foda(?:-se)?|fode(?:-te)?|cabr[ãa]o|puta|putas|filho da puta|foda|porra|cona|badalhoc[oa]|corno|otári[oa]|fuck(?:ing|ed|er)?|shit|bitch|asshole|bastard|cunt|dick(?:head)?|motherfucker)\b",
    )
    .unwrap()
});

/// Substitui palavras ofensivas por asteriscos do mesmo comprimento
/// (mantém a cadência da frase sem exibir o palavrão).
pub fn mask_profanity(text: &str) -> String {
    PROFANITY_RE
        .replace_all(text, |c: &regex::Captures| "*".repeat(c[0].chars().count()))
        .into_owned()
}

/// Limpeza de legenda/transcrição: censura DLP (PII) + máscara de palavrões.
/// Usar em todo o texto de fala difundido a outros participantes.
pub fn clean_caption(text: &str) -> String {
    mask_profanity(&censor(text))
}

/// Redige informações sensíveis de um texto.
pub fn censor(text: &str) -> String {
    let mut safe = text.to_string();

    // Censura Cartão de Crédito
    safe = CREDIT_CARD_RE
        .replace_all(&safe, "[CARTÃO DE CRÉDITO BLOQUEADO PELO DLP]")
        .to_string();

    // Censura NIF
    safe = NIF_RE
        .replace_all(&safe, "[NIF BLOQUEADO PELO DLP]")
        .to_string();

    // Censura Chaves API
    safe = API_KEY_RE
        .replace_all(&safe, "[CHAVE API CENSURADA]")
        .to_string();

    safe
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mask_profanity() {
        assert_eq!(mask_profanity("isto é uma merda pegada"), "isto é uma ***** pegada");
        assert_eq!(mask_profanity("that is pure shit man"), "that is pure **** man");
        // Conservador: só palavras isoladas (fronteira \b) — compostos e
        // substrings inocentes NÃO são apanhados (evita falsos positivos).
        assert_eq!(mask_profanity("that is bullshit"), "that is bullshit");
        assert_eq!(mask_profanity("a pucará e o cornaça"), "a pucará e o cornaça");
        // Case-insensitive:
        assert_eq!(mask_profanity("MERDA total"), "***** total");
    }

    #[test]
    fn test_clean_caption_pii_and_profanity() {
        let out = clean_caption("paga com o cartão 1234 5678 1234 5678 seu merda");
        assert!(out.contains("[CARTÃO DE CRÉDITO BLOQUEADO PELO DLP]"));
        assert!(out.contains("*****") && !out.contains("merda"));
    }

    #[test]
    fn test_censor_credit_card() {
        let input = "O meu cartão é 1234 5678 1234 5678, usa-o bem.";
        let output = censor(input);
        assert_eq!(
            output,
            "O meu cartão é [CARTÃO DE CRÉDITO BLOQUEADO PELO DLP], usa-o bem."
        );
    }

    #[test]
    fn test_censor_nif() {
        let input = "Fatura para o NIF 234567890 por favor.";
        let output = censor(input);
        assert_eq!(
            output,
            "Fatura para o NIF [NIF BLOQUEADO PELO DLP] por favor."
        );
    }

    #[test]
    fn test_censor_api_key() {
        let input = "A chave secreta é sk-abcdef1234567890abcdef1234567890 não partilhes.";
        let output = censor(input);
        assert_eq!(
            output,
            "A chave secreta é [CHAVE API CENSURADA] não partilhes."
        );
    }

    #[test]
    fn test_no_censorship_needed() {
        let input = "Olá, tudo bem? 12345";
        let output = censor(input);
        assert_eq!(output, "Olá, tudo bem? 12345");
    }
}

File: ai-worker/dlp.py
"""DLP (redação de dados sensíveis) — espelha server/src/dlp.rs.

Duplicado de propósito: o worker é Python e o servidor é Rust, sem lib
partilhada entre os dois. Os PADRÕES têm de se manter em sincronia à mão —
qualquer mudança num lado pede a mesma mudança no outro.

Sem isto, o `transcribe_worker.py` escrevia a transcrição de uma gravação
directamente na base de dados a partir do áudio bruto, sem nunca passar
pelo DLP: um cartão de crédito dito durante a reunião ficava gravado tal
e qual (achado ao alinhar com o OWASP Top 10 para LLM — LLM06, divulgação
de informação sensível).
"""
import re

CREDIT_CARD_RE = re.compile(r"(?:\d[ -]*?){13,16}")
NIF_RE = re.compile(r"\b[1-9]\d{8}\b")
API_KEY_RE = re.compile(r"sk-[a-zA-Z0-9]{32}")

PROFANITY_RE = re.compile(
    r"\b(?:merda|caralho|foda(?:-se)?|fode(?:-te)?|cabr[ãa]o|puta|putas|"
    r"filho da puta|foda|porra|cona|badalhoc[oa]|corno|otári[oa]|"
    r"fuck(?:ing|ed|er)?|shit|bitch|asshole|bastard|cunt|dick(?:head)?|"
    r"motherfucker)\b",
    re.IGNORECASE,
)


def mask_profanity(text: str) -> str:
    """Substitui palavras ofensivas por asteriscos do mesmo comprimento."""
    return PROFANITY_RE.sub(lambda m: "*" * len(m.group(0)), text)


def censor(text: str) -> str:
    """Redige cartão de crédito, NIF e chaves de API."""
    text = CREDIT_CARD_RE.sub("[CARTÃO DE CRÉDITO BLOQUEADO PELO DLP]", text)
    text = NIF_RE.sub("[NIF BLOQUEADO PELO DLP]", text)
    text = API_KEY_RE.sub("[CHAVE API CENSURADA]", text)
    return text


def clean_caption(text: str) -> str:
    """Limpeza completa: DLP (PII) + máscara de palavrões."""
    return mask_profanity(censor(text))


if __name__ == "__main__":
    # Sem pytest no repo (nenhum ficheiro `test_*.py` existe) — casos
    # espelhados dos testes de `dlp.rs`, corridos com `python3 dlp.py`.
    assert mask_profanity("isto é uma merda pegada") == "isto é uma ***** pegada"
    assert mask_profanity("that is pure shit man") == "that is pure **** man"
    assert mask_profanity("that is bullshit") == "that is bullshit"
    assert mask_profanity("MERDA total") == "***** total"
    out = clean_caption("paga com o cartão 1234 5678 1234 5678 seu merda")
    assert "[CARTÃO DE CRÉDITO BLOQUEADO PELO DLP]" in out
    assert "*****" in out and "merda" not in out
    assert censor("O meu cartão é 1234 5678 1234 5678, usa-o bem.") == (
        "O meu cartão é [CARTÃO DE CRÉDITO BLOQUEADO PELO DLP], usa-o bem."
    )
    assert censor("Fatura para o NIF 234567890 por favor.") == (
        "Fatura para o NIF [NIF BLOQUEADO PELO DLP] por favor."
    )
    assert censor(
        "A chave secreta é sk-abcdef1234567890abcdef1234567890 não partilhes."
    ) == "A chave secreta é [CHAVE API CENSURADA] não partilhes."
    assert censor("Olá, tudo bem? 12345") == "Olá, tudo bem? 12345"
    print("dlp.py: todos os testes passaram")
